File: skills/cp-solver/references/implementations/suffix-array/suffix_array.py
"""
Suffix Array + LCP Array
Complexity: O(N log N) build suffix array, O(N) build LCP
Dùng cho: string matching, longest common substring, pattern queries

sa[i]  = starting index của suffix thứ i theo thứ tự lexicographic
lcp[i] = LCP của sa[i] và sa[i-1]
"""
import sys
input = sys.stdin.readline


def build_suffix_array(s):
    """
    Build suffix array sử dụng prefix doubling — O(N log² N).
    Đủ nhanh cho N ≤ 10^5 trong Python.
    Trả về sa: list length N, sa[i] = starting index của suffix thứ i
    """
    n = len(s)
    sa = sorted(range(n), key=lambda i: s[i:])  # Initial sort O(N² log N) cho nhỏ

    # Prefix doubling O(N log² N)
    rank = [0] * n
    tmp  = [0] * n

    # Khởi tạo rank theo ký tự đầu
    rank = [ord(c) for c in s]

    k = 1
    while k < n:
        def key(i):
            return (rank[i], rank[i + k] if i + k < n else -1)

        sa = sorted(range(n), key=key)

        tmp[sa[0]] = 0
        for i in range(1, n):
            tmp[sa[i]] = tmp[sa[i-1]]
            if key(sa[i]) != key(sa[i-1]):
                tmp[sa[i]] += 1
        rank = tmp[:]

        if rank[sa[n-1]] == n - 1:
            break  # All ranks distinct, done
        k *= 2

    return sa


def build_lcp_array(s, sa):
    """
    Build LCP array sử dụng Kasai's algorithm — O(N).
    lcp[i] = LCP của suffix sa[i] và sa[i-1] (với lcp[0] = 0)
    """
    n = len(s)
    rank = [0] * n
    for i, idx in enumerate(sa):
        rank[idx] = i

    lcp = [0] * n
    h = 0
    for i in range(n):
        if rank[i] > 0:
            j = sa[rank[i] - 1]
            while i + h < n and j + h < n and s[i+h] == s[j+h]:
                h += 1
            lcp[rank[i]] = h
            if h > 0:
                h -= 1

    return lcp


def find_pattern(s, pattern, sa):
    """
    Tìm tất cả vị trí xuất hiện của pattern trong s — O(M log N).
    Trả về list các starting indices (sorted).
    """
    import bisect
    n, m = len(s), len(pattern)

    lo = bisect.bisect_left(sa, 0, key=lambda i: s[i:i+m] >= pattern)
    hi = bisect.bisect_right(sa, 0, key=lambda i: s[i:i+m] <= pattern)

    # Dùng binary search thủ công vì tinh tế hơn
    def lower():
        l, r = 0, n
        while l < r:
            mid = (l + r) // 2
            if s[sa[mid]:sa[mid]+m] < pattern:
                l = mid + 1
            else:
                r = mid
        return l

    def upper():
        l, r = 0, n
        while l < r:
            mid = (l + r) // 2
            if s[sa[mid]:sa[mid]+m] <= pattern:
                l = mid + 1
            else:
                r = mid
        return l

    lo, hi = lower(), upper()
    return sorted(sa[lo:hi])


def longest_common_substring(s, t):
    """Tìm LCS của 2 xâu s và t — O((N+M) log(N+M))."""
    combined = s + '#' + t   # '#' nhỏ hơn mọi ký tự thường
    n, m = len(s), len(t)
    N = len(combined)

    sa  = build_suffix_array(combined)
    lcp = build_lcp_array(combined, sa)

    best = 0
    for i in range(1, N):
        # LCP giữa sa[i] và sa[i-1]
        # Nếu hai suffix từ hai xâu khác nhau
        p = sa[i-1]
        q = sa[i]
        if (p < n) != (q < n):   # Một trong s, một trong t
            best = max(best, lcp[i])

    return best


# ===== USAGE EXAMPLE =====
def solve():
    s = input().strip()
    sa = build_suffix_array(s)
    lcp = build_lcp_array(s, sa)

    # In suffix array
    for i in sa:
        print(i, s[i:])

    # LCP
    print("LCP:", lcp)


if __name__ == '__main__':
    solve()

File: skills/cp-solver/references/implementations/suffix-array/suffix_array.cpp
// Suffix Array (O(N log² N) prefix doubling) + LCP Array (Kasai O(N))
// Dùng cho: pattern matching, longest common substring, distinct substrings

#include <bits/stdc++.h>
using namespace std;

// Build suffix array — O(N log² N)
vector<int> build_suffix_array(const string &s) {
  int n = s.size();
  vector<int> sa(n), rank_(n), tmp(n);

  iota(sa.begin(), sa.end(), 0);
  for (int i = 0; i < n; i++)
    rank_[i] = s[i];

  for (int k = 1; k < n; k <<= 1) {
    auto cmp = [&](int a, int b) {
      if (rank_[a] != rank_[b])
        return rank_[a] < rank_[b];
      int ra = (a + k < n) ? rank_[a + k] : -1;
      int rb = (b + k < n) ? rank_[b + k] : -1;
      return ra < rb;
    };
    sort(sa.begin(), sa.end(), cmp);

    tmp[sa[0]] = 0;
    for (int i = 1; i < n; i++)
      tmp[sa[i]] = tmp[sa[i - 1]] + (cmp(sa[i - 1], sa[i]) ? 1 : 0);
    rank_ = tmp;

    if (rank_[sa[n - 1]] == n - 1)
      break;
  }
  return sa;
}

// Build LCP array — Kasai O(N)
vector<int> build_lcp(const string &s, const vector<int> &sa) {
  int n = s.size();
  vector<int> rank_(n), lcp(n, 0);
  for (int i = 0; i < n; i++)
    rank_[sa[i]] = i;

  int h = 0;
  for (int i = 0; i < n; i++) {
    if (rank_[i] > 0) {
      int j = sa[rank_[i] - 1];
      while (i + h < n && j + h < n && s[i + h] == s[j + h])
        h++;
      lcp[rank_[i]] = h;
      if (h > 0)
        h--;
    }
  }
  return lcp;
}

// Pattern matching — tìm tất cả vị trí xuất hiện của pattern trong text — O(M
// log N)
pair<int, int> find_pattern(const string &s, const string &pattern,
                            const vector<int> &sa) {
  int n = sa.size(), m = pattern.size();

  int lo = 0, hi = n;
  // Lower bound: suffix >= pattern
  while (lo < hi) {
    int mid = (lo + hi) / 2;
    if (s.substr(sa[mid], m) < pattern)
      lo = mid + 1;
    else
      hi = mid;
  }
  int left = lo;

  lo = 0;
  hi = n;
  // Upper bound: suffix > pattern
  while (lo < hi) {
    int mid = (lo + hi) / 2;
    if (s.substr(sa[mid], m) <= pattern)
      lo = mid + 1;
    else
      hi = mid;
  }
  int right = lo;

  return {left, right}; // sa[left..right) chứa các occurrences
}

// Longest Common Substring của s và t — O((N+M) log(N+M))
string lcs(const string &s, const string &t) {
  string combined = s + "#" + t; // '#' < 'a'
  int n = s.size(), m = t.size(), N = combined.size();

  auto sa = build_suffix_array(combined);
  auto lcp = build_lcp(combined, sa);

  int best_len = 0, best_pos = -1;
  for (int i = 1; i < N; i++) {
    int p = sa[i - 1], q = sa[i];
    // Nếu hai suffix từ hai xâu khác nhau
    if ((p < n) != (q < n)) {
      if (lcp[i] > best_len) {
        best_len = lcp[i];
        best_pos = sa[i];
      }
    }
  }

  if (best_pos == -1)
    return "";
  if (best_pos < n)
    return s.substr(best_pos, best_len);
  return t.substr(best_pos - n - 1, best_len);
}

// Đếm số xâu con phân biệt — O(N log N)
long long count_distinct_substrings(const string &s) {
  int n = s.size();
  auto sa = build_suffix_array(s);
  auto lcp = build_lcp(s, sa);

  long long total = (long long)n * (n + 1) / 2; // Tổng tất cả xâu con
  for (int i = 1; i < n; i++)
    total -= lcp[i]; // Trừ trùng lặp
  return total;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  string s;
  cin >> s;

  auto sa = build_suffix_array(s);
  auto lcp = build_lcp(s, sa);

  // In suffix array
  for (int i : sa)
    cout << i << " ";
  cout << "\n";

  // LCP
  for (int x : lcp)
    cout << x << " ";
  cout << "\n";

  // Distinct substrings
  cout << count_distinct_substrings(s) << "\n";

  return 0;
}
